Add command-line options

jzohrab opened the pull request:
Building on my prior pull requests :-) this PR adds options to index.js so that I can control the export better.  E.g.:

```
$ node index.js -p tregua -p miedo -o ~/Downloads/some/folder --datesubfolder false --nospaces
```

See the README updates.

ps - I don't know if anyone cares about these updates I'm making, but figure I'll PR them to here as the changes are useful for me, at least!

Cheers, jz

jzohrab committed:
Add yargs package and some initial option ideas.

jzohrab committed:
Use args to determine output folder.

jzohrab committed:
Use args to get file patterns.

jzohrab committed:
Handle spaces in filename.

jzohrab committed:
Extract function.

jzohrab committed:
Refactor.

jzohrab committed:
Create all directories needed before exporting.

jzohrab committed:
README updates.

jzohrab committed:
Running with no args should behave as before.

hagent commented on index.js:
```diff
@@ -1,10 +1,45 @@
 const { exec } = require("child_process");
 const sanitize = require("sanitize-filename");
 
+const yargs = require('yargs');
```
`yargs` should be added to package.json

hagent commented on index.js:
```diff
@@ -96,54 +135,146 @@ original error: ${e}`);
+}
+
+// Default: return all files.
+let patterns = []
```
nitpicking: I would move that logic to `main` to have on the top level only function call

jzohrab replied:
Yes, that's better, thanks.

hagent commented on index.js:
```diff
@@ -96,54 +135,146 @@ original error: ${e}`);
+  }));
+  let filteredPodcasts = filterPodcasts(podcasts, filepatterns);
+
+  // Weirdly, there are some podcasts that are duplicates ...  i.e.,
```
I don't have duplicates, but in theory it can be that app for example cached the same podcast 2 times with different uuid, should not really be a problem, even if there are duplicates they will be exported 2 times, if you want to debug it you can export such duplicates with some special name and check whether it's indeed the same podcast episode, but anyway I think that comment can be removed

jzohrab replied:
Funnily enough, there were dups in mine!  I didn't look into the podcast db, but when I ran it, the code said "already have X, skipping", even on a fresh export.  Since that looked odd (to me as the user), I left in the check.

hagent commented on index.js:
```diff
@@ -96,54 +135,146 @@ original error: ${e}`);
+  */
+  // Since this causes some strange messages to appear during output
+  // to a new output directory, delete the dups by keying on filename.
+  const temp = {}
+  // console.log('removing dups');
+  filteredPodcasts.forEach(p => temp[p.exportFileName] = p);
```
I'm still not sure why duplication is a problem, but regarding this uniq values I think Object.fromEntries might be handy
```suggestion
  const temp = Object.fromEntries(filteredPodcasts.map(podcast => [podcast.exportFileName, podcast]))
```
also maybe instead of `temp` some name like `uniqByFilename` could work better

jzohrab replied:
sure, np.

jzohrab replied:
Re "I'm still not sure why duplication is a problem" - just because it looks weird on output.  :-)  e.g., in my case:

```
$ node index.js CHURCHI
Exporting 2 of 464
344DA728-FE9D-40CC-92F6-2E27260EF79E.mp3 -> La Segunda Guerra Mundial (EPT)/Episodio 65. Churchill, Stalin, Roosevelt Y Sus Pr.mp3
2D78D27E-AEF7-48E2-BCA8-E270145E7BDD.mp3 -> La Segunda Guerra Mundial (EPT)/Episodio 65. Churchill, Stalin, Roosevelt Y Sus Pr.mp3
```

It's not a _problem_, it just looks/feels weird.

jzohrab replied:
Changed to `const uniqByFilename = Object.fromEntries(filteredPodcasts.map(p => [p.exportFileName, p]));`, thanks.  :-)

jzohrab commented:
@hagent - I'll wait until #6 is reviewed and merged before updating this one, as this code builds on that one.  Thanks! z

jzohrab committed:
Move pattern arg to main per code review.

jzohrab committed:
Semicolon for linter.

jzohrab commented:
Rebased the branch off of upstream/main, and re-pushing now.  I _believe_ I haven't lost any of the prior feedback in the cherry-picking of commits.  Now fixing linter issues.

jzohrab committed:
Commit package-lock.json per github ci failure.

Per https://github.com/hagent/export-macos-podcasts/runs/6562193825,
ran 'npm install' and committed lock file.

jzohrab committed:
Fix linter errors.

jzohrab commented:
I've resolved all of the conversations, thanks for the feedback.

@hagent , I'm currently getting the following linter errors:

```
Error:   1:7   error  Expected a line break after this opening brace   object-curly-newline
Error:   1:59  error  Expected a line break before this closing brace  object-curly-newline
Error:   9:7   error  Use object destructuring                         prefer-destructuring
```

I didn't touch line 1, and the note about preferring object destructuring is odd -- at least, every example I've seen using yargs, or have written myself, has followed the syntax and usage I've used here.  Thoughts?

In the meantime, I've tried out the code with the various flags to ensure it still works -- lots of conflicts during rebasing (???), and it seems fine.  (e.g. `node index.js -p CHURCH --nospaces -o .`)
